src/main/java/com/rtb/projectmanagementtool/project/ProjectData.java

```diff
@@ -0,0 +1,204 @@
+/**
+ * ProjectData.java - this file implements the Project class, which contains details about a
+ * particular project, including the members and tasks within it.
+ */
+package com.rtb.projectmanagementtool.project;
+
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.Key;
+import com.google.appengine.api.datastore.KeyFactory;
+import com.google.gson.Gson;
+import com.google.gson.reflect.TypeToken;
+import java.lang.reflect.Type;
+import java.util.HashMap;
+import java.util.HashSet;
+
+enum User {
+  CREATOR,
+  ADMIN,
+  REGULAR;
+}
+
+public class ProjectData {
+  /**
+   * constant variables used to specify the names of each property of the Project entity. Used to
+   * minimize redundancy and also make renaming properties easy if needed
+   */
+  private final String PROPERTY_NAME = "name";
+
+  private final String PROPERTY_DESCRIPTION = "description";
+  private final String PROPERTY_USER_IDS = "userIds";
+  private final String PROPERTY_TASK_IDS = "taskIds";
+
+  private DatastoreService datastore;
+
+  private long id;
+  private String name;
+  private String description;
+  private HashMap<Long, User> userIds;
+  private HashSet<Long> taskIds;
+
+  /**
+   * Class constructor with the minimum requirements for creating a project.
+   *
+   * @param name the name of the project
+   * @param description the description of the project
+   * @param creatorId the id of the user who created the project
+   */
+  public ProjectData(String name, String description, long creatorId) {
+    datastore = DatastoreServiceFactory.getDatastoreService();
+    this.id = datastore.put(new Entity("Project")).getId();
+    this.name = name;
+    this.description = description;
+
+    this.userIds = new HashMap<Long, User>();
+    this.userIds.put(creatorId, User.CREATOR);
+
+    this.taskIds = new HashSet<Long>();
+    toEntity();
+  }
+
+  /**
+   * Class constructor to create Project from existing datastore entity
+   *
+   * @param entity the datastore entity to parse object from
+   */
+  public ProjectData(Entity entity) {
+    datastore = DatastoreServiceFactory.getDatastoreService();
+
+    this.id = (Long) entity.getKey().getId();
+    this.name = (String) entity.getProperty(PROPERTY_NAME);
+    this.description = (String) entity.getProperty(PROPERTY_DESCRIPTION);
+
+    Gson gson = new Gson();
+
+    // sets the type of containers to store parsed JSON in
+    Type mapType = new TypeToken<HashMap<Long, User>>() {}.getType();
+    Type setType = new TypeToken<HashSet<Long>>() {}.getType();
+
+    this.userIds = gson.fromJson((String) entity.getProperty(PROPERTY_USER_IDS), mapType);
+    this.taskIds = gson.fromJson((String) entity.getProperty(PROPERTY_TASK_IDS), setType);
+  }
+
+  /** Saves the object as a datastore entity. */
+  private void toEntity() {
```

Comment:
`toSomething` naming conventions usually implies conversion and returning the converted value.  think about `toString` that would return a string conversion of an object, etc..

Reply:
Could this method (or some other form of it) be used to not just create new entitites but also to update existing ones? 
The problem with returning an entity here is that when we try to add it to the datastore in ProjectController, a whole new entity will be created, and that isn't ideal if we just want to update the entity's properties.

Reply:
> The problem with returning an entity here is that when we try to add it to the datastore in ProjectController, a whole new entity will be created, and that isn't ideal if we just want to update the entity's properties.

I don't think we necessarily have to push Entities directly into the Datastore as soon as they're created.  Entities, being objects themselves, can be passed around, stored, or operated on later on.  This will allow us to better separate your Data Model from your operations.

I think the question you're thinking about here illustrates a good use case for why you'd want to separate the Model from your Controller.  The entity itself (ProjectData), shouldn't really care if it's being added or being updated.  All it cares about is how its fields are filled in, and how it shares it fields to other consumers.

What you do with that Data is someone else's (Controller's) responsibilities.  That someone else is the one that needs to keep track of the difference between an insert and an update.

Reply:
Okay that makes sense